Design document for user customization and persistence (#2241)
scality/metalk8s: docs/developer/architecture/user_customization_and_persistence.rst

editable configurations in a MetalK8s cluster based on the above requirements
for the following reasons:

* Ability to support Update operations on ConfigMap's with CLI and UI easily
  using our already existing python kubernetes module.
* Guarantee of adaptability and ease of changing the design and implementation
  in cases were customer needs evolve rapidly.
* ConfigMaps are stored in the :term:`etcd` database which is generally being
  backed up. This ensures that user settings cannot be lost easily.

How it works
~~~~~~~~~~~~

Before Bootstrap, :file:`defaults.yaml` will be pre-filled with default service
settings. Service pods and deployments will be configured to either
consume configuration data directly from ConfigMaps or fallback to
the default values if the service specific ConfigMap was not found.

**Using Salt states**

Once a ConfigMap is updated by the user (say a user changes the number of
replicas for Prometheus deployments to a new value), then perform the
following actions:

  - Apply a Salt states that reads the ConfigMap object, validates the schema
    and checks the new values passed.

Storage format
~~~~~~~~~~~~~~

> A sample ConfigMap can be defined with the following fields.

Reviewer: Why this format? Explain in the design, compare with flat KV approach and why we prefer this one.

Author: Done

        kind: GrafanaConfig
        spec:
          deployment:
            replicas: 2

Non-goals
~~~~~~~~~

This section contains requirements stated above which the current design choice
does not cater for and will be addressed later:

- Persisting newly added Grafana dashboards or new Grafana datasources
  especially for modifications added via the Grafana UI cannot be stored in
  ConfigMaps and hence will be catered for later.

- As stated in the requirements, adding and editing Prometheus alert rules
  is also not covered by the chosen design choice and will need to be addressed
  differently.

Rejected design choices
~~~~~~~~~~~~~~~~~~~~~~~

Consul KV vs ConfigMap
~~~~~~~~~~~~~~~~~~~~~~

This approach offers a full fledge KV store with a /kv endpoint which allows
CRUD operations on all KV data stored in it.
Consul KV also allows access to past versions of objects and has an optimistic
concurrency when manipulating multiple objects.

Note that, Consul KV store was rejected because managing operations such as
performing full backups, system restores for a full fledged KV system
requires time and much more efforts than the ConfigMap approach.

> Operator (Custom Controller) vs ConfigMap

Reviewer: `vs. Salt` - as Teddy said, Operator is not related to a given storage medium, and could work with plain CM

User Customization & Persistence
================================

Context
-------

.. todo::

   This section will be handled by the requirements PR.

Design Choices
--------------

:term:`ConfigMap` is chosen as a unified data access and storage media for user
editable configurations in a MetalK8s cluster based on the above requirements
for the following reasons:

* Ability to support Update operations on ConfigMap's with CLI and UI easily
  using our already existing python kubernetes module.
* Guarantee of adaptability and ease of changing the design and implementation
  in cases were customer needs evolve rapidly.
* ConfigMaps stored in the :term:`etcd` database which is generally being
  backed up. This ensures that user settings cannot be lost easily.

How it works
~~~~~~~~~~~~

During Bootstrap, pillar values for the various services will be pre-filled
with default values. Service pods and deployments will be configured to either
consume configuration data directly from their ConfigMap or fallback to

> the default pillar values if the service specific ConfigMap was not found.

Reviewer 2: What if we did this differently?
None of these 'user configurable' settings (at this level) should influence how the K8s cluster behaves: this should be dictated by the `BootstrapConfiguration`. As such during bootstrap, at some point *before we need any of these settings*, the K8s cluster is fully available.

Which means that at this point, we could create all these `ConfigMap`s, filled with whatever default values we choose.

At this point, we can consume them while deploying various services without any more 'defaults' defined anywhere.

Author: True! after a couple of discussions, an alternative we considered and updated the design docs was keeping these 'defaults' in our `default.yaml` file and then doing something like ```{%- from "metalk8s/map.jinja" import addons with context %}```. This meant in our add-ons charts we had to implement some conditional to read from the Configmaps or fall back to these defined defaults(just like pillar pre-filling idea but cleaner).

Thanks to your approach, it appears we do not even need to declare defaults in `defaults.yaml`.  Waiting to hear from others what they think.

Reviewer 3: Like I told you Monday 3 choice:
- We have `defaults.yaml` containing this default number of replica
- we create the custos CM and "pre-fill" them **if they do not exists**
- We just put default value directly in the `sls` file

Personally suggest the second one as it will be much simpler for the user to change them if CM already exists and already contain value we want

Reviewer: Option 2 looks good to me as well. A simple `kubectl delete cm ..` followed by `salt-run state.orch metalk8s.deployed` to restore defaults is acceptable I guess.

**Using Salt states**

Once a ConfigMap is updated by the user (say a user changes the number of
replicas for Prometheus deployments to a new value), then perform the
following actions:

  - Apply a Salt states that reads the ConfigMap object, validates the schema
    and checks the new values passed.

  - Finally, we make sure that the updated values are picked up by their
    respective consumers (this might require Pod restarts for changes to take

>     effect).

Reviewer 2: The fact that Pods may be restarted is always the case in K8s, not sure this should be mentioned. In the end, Salt (MetalK8s) is not restarting any Pods. This should all be managed by the `Deployment`/`StatefulSet`/`DaemonSet` controllers because Salt updates some high-level object(s).

Author: removed

Note that, Salt states are used to sync data and update consumers
of new configurations changes mainly because of the minimum effort it takes to

> setup this flow but the K8s Operator pattern could be use to replace

Reviewer 2: ```suggestion
setup this flow but the K8s Operator pattern could be used to replace
```

> configuration synchronization between user defined configurations and

Reviewer 2: `... could be used to synchronize configuration between user-defined settings and the applications.`

consumers.

Non-goals
~~~~~~~~~

- Persisting newly added Grafana dashboards or new Grafana datasources
  especially for modifications added via the Grafana UI cannot be stored in
  ConfigMaps.

>   To handle this particular user case, we need to configure Grafana deployments

Reviewer 2: ```suggestion
  To handle this particular use-case, we need to configure Grafana deployments
```

>   to use persistent volume claims and then document how to create volumes that

Reviewer 2: Here you're prescribing *how* a certain feature *not* covered by this design ought to be implemented. That's not OK. Indeed, this document should describe what it *won't* cover, but whether this other feature is implemented using PVs or some other mechanism is to be determined later/elsewhere.

Author: True, I will move it to non-goals and clearly state that the current design does not cater for this. No need to mention an implementation choice.

  will handle the persistence of these settings across Pod restarts.

- ConfigMaps as used in the above design for storing and persisting user

>   settings need periodic backups for recovery purposes. This implies that

Reviewer 2: I get what you mean but I don't like how it's phrased. We don't need to put backups in place to store these user settings and such. Instead, using `ConfigMap`s (or other K8s object types) has as a benefit that K8s cluster backups include these settings. The fact that the K8s `etcd` cluster needs to be backed up is not driven by this feature.

Rejected design choices
~~~~~~~~~~~~~~~~~~~~~~~

Consul KV vs ConfigMap
~~~~~~~~~~~~~~~~~~~~~~

This approach offers a full fledge KV store with a /kv endpoint which allows
CRUD operations on all KV data stored in it.
Consul KV also allows access to past versions of objects and has an optimistic
concurrency when manipulating multiple objects.

Note that, Consul KV store was rejected because managing operations such as
performing full backups, system restores for a full fledged KV system
requires time and much more efforts than the ConfigMap approach.

Operator (Custom Controller) vs ConfigMap
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Operators are useful in that, they provide self-healing functionalities on a
reactive basis. When a user changes a given configuration, it is easy to
reconcile and apply these changes to the in-cluster objects.

The Operator approach was rejected because it is much more complex, requires
much more effort to realize and there is no real need for applying changes
using this method because configuration changes are not frequent
(for a typical MetalK8s admin, changing the number of replicas for a given

> deployment could happen once in 3 months) as such, having an operator watch for

Reviewer 2: Or less.

object changes is not significant and not very useful at this point in time.

Implementation Details
----------------------

Storage format
~~~~~~~~~~~~~~

A sample ConfigMap can be defined with the following fields.

An example of such a ConfigMap:

.. code-block:: yaml

    apiVersion: v1
    kind: ConfigMap
    metadata:
      namespace: <namespace>
      name: <config-name>
    data:
      config.yaml: |-
        apiVersion: <object-version>
        kind: <kind>
        spec:
          <key>: <values>

**Use case 1:**

Configure and store the number of replicas for service specific Deployments
found in the `metalk8s-monitoring` namespace using the ConfigMap format.

.. code-block:: yaml

    apiVersion: v1
    kind: ConfigMap
    metadata:
      namespace: metalk8s-monitoring
      name: metalk8s-grafana-userconfig
    data:
      config.yaml: |-
        apiVersion: metalk8s.scality.com/v1alpha1

>         kind: GrafanaUserConfig

Reviewer 2: Not sure I like the `GrafanaUserConfig` name. We're not configuring Grafana users here... `GrafanaCustomizations`? `GrafanaConfig`? ...

Author: Picked the latter!

        spec:

>           replicas: 2

Reviewer 2: This is a low-level setting at top level in the config. Not sure that's great.
```yaml
spec:
  deployment:
    replicas: 2
```
(then other `deployment`-related settings can go there as well)

Author: ok

Iteration 1
~~~~~~~~~~~

- Define and deploy new ConfigMap stores that will hold user configurations
  as listed in the requirements
- Template and render Deployment and Pod manifests that will make use of
  this user customizable configurations using pillar values
- Document how to change user configurations using kubectl

> - Document how to add Prometheus alerting rules by editing the Prometheus rule

Reviewer 2: I don't see how this is in scope of this design?

Author: fixed

  CRD.

Iteration 2
~~~~~~~~~~~

- Provide a CLI tool for changing any of the user configurations:

    - Count of replicas for chosen Deployments (Prometheus)
    - Updating a Dex authentication connector (OpenLDAP, AD and
      staticUser store)
    - Updating the Alertmanager notification configuration

- Provide a UI interface for adding, updating and deleting service specific
  configurations for example Dex-LDAP  connector integration.
- Provide a UI interface for listing MetalK8s available/supported Dex
  authentication Connectors
- Provide a UI interface for enabling or disabling Dex authentication
  connectors (LDAP, Active Directory, StaticUser store)
- Provide a UI interface for changing the number of replicas for a chosen set
  of MetalK8s deployments (Prometheus, ....)
- Add a UI interface for listing Alertmanager notification systems MetalK8s

>   will support (Slack, email, hipchat)

Reviewer 2: HipChat is dead, no need to list it anywhere.

User Customization & Persistence
================================

Context
-------

.. todo::

   This section will be handled by the requirements PR.

Design Choices

> --------------

Reviewer: This design only discusses the choice of storage (method and media). What about the concerns of translating these user configurations into services configurations? Are the services supposed to know about our user configs and parse them? Are we (Salt) deriving actual service configs from our user configs? Why Salt and not a K8s Operator?

Author: I added a `How it works section` that takes care of the questions raised here. This might need validation from You

Reviewer: Good content, though I'd rather state this in the _Design Choices_ section. In addition, there seems to be two implementation alternatives (Salt vs. Operator). Since we apparently decide to use Salt (and not an Operator), why not show Salt as the design + Operator as a rejected option?

> :term:`ConfigMap` KV(Key-Value) store is chosen as a unified data access and

Reviewer: I'm uncomfortable with this focus on "key-value" nature of a `ConfigMap` - does it imply we will use a single `ConfigMap` with a key for each "service" to be configured? I'm not sure that's what we want actually, for multiple reasons (which may be adjusted after we get the Requirements btw):

- CMs are namespace-scoped, and so are most of the services we deploy with MetalK8s. Wouldn't it make more sense if we keep a configuration CM next to what it configures? E.g. `metalk8s-auth-userconfig` in the `metalk8s-auth` namespace?
- Editing a single CM with many keys is less user-friendly than having smaller CMs per service, each with a single key. Although we plan on providing better interfaces through CLI and UI, it would make sense to make the underlying format as friendly as possible (this will make a first iteration with no CLI/GUI much more useful)

Author: Done, removed the focus on `key-value` nature

storage media for user editable configurations in a MetalK8s cluster based on
the above requirements for the following reasons:

* The limited number of read/write operations even for multi-node setups means
  that concurrency issues on CRUD operations will not lag even if we do not

>   implement cross-app locking.

Reviewer: "concurrency issues [...] will not lag" - not sure what you meant?

In any case, is it important to mention in the design? We're not looking for a database, we're looking for means of configuration. As a matter of fact, we **do have** a database (`etcd`), and a nice schema already in place that we can use (the K8s API resources, including `ConfigMap`s) for such configurations.

The concurrency matters on configuration of cluster-wide services are, as you suggest here, not a problem. Let's move this point to the "non-goals" in the Requirements (cc @thomasdanan), and leave it aside.

We need our design choices to clearly explain **why** we chose this approach according to our "goals", not to our "non-goals", IMO.

Author: Done

> * Ability to support CRUD operations on ConfigMap's with CLI and UI easily

Reviewer: Why this focus on CRUD? I'm actually thinking, we only want the users to perform **U**pdate operations on those objects, since MetalK8s is responsible for **C**reating and **D**eleting those objects - so the fact that all CRUD ops are supported could be a problem (luckily, we can simply have a `Role` which only allows `patch` on the right objects).

Author: ok

  using our already existing python kubernetes module.
* Guarantee of adaptability and ease of changing the design and implementation

>   in cases were customer needs evolve rapidly(modular design nature)

Reviewer: What is modular in this design? Especially with the "single ConfigMap KVS" approach.

* Ease of operation and a guarantee of no downtime in a MetalK8s cluster

>   especially when configuration changes are made on the fly.

Reviewer: How does using ConfigMaps handle "no-downtime" guarantees?

Reviewer: > changes are made on the fly.

WDYM? Do you expect a change to impact the target service right away? Or not?
"made on the fly" would mean something like `kubectl edit cm ...` or something else?

Author: Well, this was supposed to mean; the fact that we are making use of ConfigMaps, I think once a user updates a given configuration, we do not require Pod restarts for the specific services to pick up the changes. no?

Rejected design choices
~~~~~~~~~~~~~~~~~~~~~~~

Consul KV store
~~~~~~~~~~~~~~~

This approach offers a full fledge KV store with a /kv endpoint which allows
CRUD operations on all KV data stored in it.
Consul KV also allows access to past versions of objects and has an optimistic
concurrency(CRUD operations will not lag with load).

Note that, Consul KV store was rejected because managing operations such as
performing full backups, system restores for a full fledged KV system
requires time and much more efforts than the ConfigMap KV store which is
simplistic and matches the requirements stated.

Implementation Details
----------------------

A sample ConfigMap KV store can be defined with the following fields. These
objects will be namespace bounded ensuring that sensitive information is

> restricted from a cluster-wide access.

Reviewer: CMs are always namespace-scoped.

Author: ok

An example of such a ConfigMap KV store:

   .. code-block:: yaml

       apiVersion: v1
       kind: ConfigMap
       metadata:
         namespace: <metalk8s-namespace>
         name: configmap-kv
       data:
         <key>: <value>
         <key>: <value>
         <key>: <value>

**Use case 1:**

Configure and store the number of replicas for all Deployments found in the
`metalk8s-monitoring` namespace using the ConfigMap KV store format.

   .. code-block:: yaml

       apiVersion: v1
       kind: ConfigMap
       metadata:
         namespace: metalk8s-monitoring
         name: <example-name>
       data:
         replicas.grafana: 2
         replicas.prometheus: 2

>          replicas.alertmanager: 1

Reviewer: This kind of format is not straightforward to manage with code, since you'll need to parse the flat keys into hierarchical structures. As discussed during the short design session, we'd rather have something more "K8s-like":

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metalk8s-monitoring
  name: metalk8s-grafana-userconfig
data:
  config.yaml: |-
    apiVersion: metalk8s.scality.com/v1alpha1
    kind: GrafanaUserConfig
    spec:
      replicas: 2
```

This also has the benefit of bearing a schema version, which we can check before inferring the underlying service configuration.

Author: Thanks for the reminder! even though i think at the end of the day, we would have littered the entire namespace with many small ConfigMap objects.

What are the chances that we could group the number of replicas specific ConfigMap for all services in the Monitoring namespace?

Reviewer 3: Not really agree with your approach here.
If we use a config map IMO we should have just "key value" like Claude did and when we will use specific CRD then we could have object like you suggest.
Also IMO we should do one configmap per "object/service" (e.g.: One configmap for auth/dex, one for grafana, one for prometheus (or maybe one for monitoring regrouping prom, grafana altermanager), one for ...)

Author: > Thanks for the reminder! even though i think at the end of the day, we would have littered the entire namespace with many small ConfigMap objects.
> 
> What are the chances that we could group the number of replicas specific ConfigMap for all services in the Monitoring namespace?

Oh, let us keep it service-specific as suggested since the configmap of one particular service might grow drastically when new customizations are included

Reviewer: > If we use a config map IMO we should have just "key value" like Claude did and when we will use specific CRD then we could have object like you suggest.

Then it means we'll need to implement the parsing logic twice? Why not stick with the same YAML style/format?

Reviewer 3: The configmap will be read from salt so if we use the "kubernetes style" that mean we need to implement it in our salt modules as we will have a configmap with several "key=string" and we will need to then load and parse the string to get the yaml we want.
If we stick with key=value we will just have to read a classic dict in salt so it's really easy.

        channel.smtp.smtp_from: 'foo@bar.com'
        channel.smtp.smtp_auth_username: 'foo@bar.com'
        channel.smtp.smtp_auth_password: 'barfoo'

        channel.slack.slack_api_url: 'https://hooks.slack.com/services/abc123'

        channel.hipchat.hipchat_auth_token: '1234556789'
        channel.hipchat.hipchat_url: 'https://hipchat.foobar.org/'

        # route.group_wait: '30s'
        # route.repeat_interval: '15m'
        # route.group_interval: '5m'
        route.receiver: 'default'

        routes.match.severity: 'email'
        routes.receiver: 'default'
        routes.match.severity: 'slack'
        routes.match.receiver: 'slack_alert'

        receivers.name: 'default'
        receivers.email_configs.to: 'foo@bar.com'

        receivers.name: 'slack_alert'
        receivers.slack_configs.channel: '#devops'
        receivers.slack_configs.send_resolved: 'true'

Iteration 1
~~~~~~~~~~~

- Define and deploy new namespaced bounded ConfigMap KV stores that will hold
  user configurations as listed in the requirements
- Template and render Deployment and Pod manifests that will make use of
  this user customizable configurations using default pillar values

> - Provide a CLI tool(Bash script) for changing any of the user configurations:

Reviewer 4: Why Bash instead of Python?
I thought we were moving away from Shell script.

Author: Very true @slaperche-scality 
I think we have an entire EPIC to handle this migration and it is debatable if we should wait or have something working(shell scripts) while waiting.

Maybe, after all, we do not even have to handle this point within this EPIC

Reviewer 4: The other EPIC is about migrating existing shellscript.

Meanwhile, for new ones that we creates from scratch, it could be a good idea to use Python from the beginning (except if it's really more costly than Bash for some reasons): would be less work for the other EPIC, while adding no extra-work to the current one (anyway you have to write a new script)

Reviewer: In my opinion, wait for the Python CLI epic (#2163) to have the foundations set in place before implementing a CLI for this. In any case, the CLI / UI implementation of those customizations is less urgent than having a framework to persist said customizations over upgrade/downgrade/... IMO, CLI / UI shouldn't be part of the first iteration.

> - Provide a CLI tool(Bash script) for changing any of the user configurations:

Reviewer: CLI tooling doesn't need to be part of the first iteration IMO - well formatted CMs are easy enough to manage both manually and in tests/scripts.

    - Count of replicas for chosen Deployments(Prometheus)
    - Adding/modifying/removing a Dex authentication connector(OpenLDAP, AD and
      staticUser store)
    - Adding/modifying/removing Alertmanager notification configuration

> Iteration 2

Reviewer: UI and CLI could both be part of iteration 2 I think, and we could add a third iteration to include pre-edit validation hooks, display of both user config and generated service configs... We'll see with the reqs.

Author: Moved UI and CLI operations to iteration 2 as suggested. Not sure about third iteration so i'm gonna have to pass!

~~~~~~~~~~~

- Provide a UI interface that allows CRUD operations on all user customizable
  settings based on the above requirements
- Provide a UI interface for adding, updating and deleting service specific
  configurations for example Dex-LDAP  connector integration.
- Provide a UI interface for listing MetalK8s available/supported Dex
  authentication Connectors
- Provide a UI interface for enabling or disabling Dex authentication
  connectors(LDAP, Active Directory, StaticUser store)
- Provide a UI interface for changing the number of replicas for a chosen set
  of MetalK8s deployments(Prometheus, ....)
- Add a UI interface for listing Alertmanager notification systems MetalK8s
  will support(Slack, email, hipchat)
- Provide a UI interface for adding, modifying and deleting Alertmanager
  configurations from the listing above

Documentation
-------------

In the Operational Guide:

* Document how to customize or change any given setting using the CLI tool

> * Document how to customize or change any given setting using the UI interface

Reviewer: OK, maybe a "common doc" for CLI / UI. But we'll also need this info next to the services we want our users to configure (e.g. show how to configure Dex in the section/page dedicated to Dex)

Test Plan
---------

Dex Static User authentication is currently covered in our test-suite and it
will make sense to cover atleast one other authentication connector with the
easiest being LDAP since we readily have access to OpenLDAP Docker images and
automating this process is possible on Scality Cloud.

CRUD Operations on the ConfigMap KV store should be tested both from the UI
test libraries and the CLI tool provided(An initial idea here is to have a
simplistic configurator.sh bash script(CLI tool) performing sets of CRUD

> operations on a sample ConfigMap KV store).

Reviewer: Standard operations to edit configmaps are not super useful to test - what we want to check is propagation of those changes down to the configured services.
E2E tests with CLI and UI are a good thing, because (I guess) those would also trigger the right Salt calls after the CM(s) update(s).

We also need to test edge-cases:
- invalid format for the user configs
- invalid values in the user configs
- lost a user config (someone deleted it)

Author: ok

User Customization & Persistence
================================

Context
-------

.. todo::

   This section will be handled by the requirements PR.

Design Choices
--------------

:term:`ConfigMap` store is chosen as a unified data access and
storage media for user editable configurations in a MetalK8s cluster based on
the above requirements for the following reasons:

* Ability to support Update operations on ConfigMap's with CLI and UI easily
  using our already existing python kubernetes module.
* Guarantee of adaptability and ease of changing the design and implementation
  in cases were customer needs evolve rapidly.

> * ConfigMap data is store in the :term:`etcd` database which is generally being

Reviewer: ```suggestion
* ConfigMaps are stored in the :term:`etcd` database which is generally being
```

  backed up. This ensures that user settings cannot be lost easily.

.. note::

   Persisting newly added Grafana dashboards or new Grafana datasources
   especially for modifications added via the Grafana UI cannot be stored in
   ConfigMaps.
   To handle this particular user case, there is a need to provision persistent
   storage volumes to handle the persistence of these settings across Pod

>    restarts.

Reviewer: True. I would create a section for "non-goals" of this design, including this point on Grafana, but also other aspects (providing a general solution for recovery after total site loss, periodic backups, ...).

Rejected design choices
~~~~~~~~~~~~~~~~~~~~~~~

Consul KV
~~~~~~~~~

This approach offers a full fledge KV store with a /kv endpoint which allows
CRUD operations on all KV data stored in it.
Consul KV also allows access to past versions of objects and has an optimistic
concurrency when manipulating multiple objects.

Note that, Consul KV store was rejected because managing operations such as
performing full backups, system restores for a full fledged KV system
requires time and much more efforts than the ConfigMap KV store which is
simplistic and matches the requirements stated.

Implementation Details
----------------------

> Storage format

Reviewer: The actual format is part of the design. It would be worth adding an explanation in the _Design Choices_ as to why we want to use the traditional versioned `apiVersion/kind` YAML style instead of flat KV style (what benefits it brings, what constraints / extra efforts it requires...).

Author: Well, we still have an open discussion about the format to use and from what I see, the flat KV style sits well with our already existing python-Kubernetes module....

~~~~~~~~~~~~~~

A sample ConfigMap store can be defined with the following fields.

An example of such a ConfigMap store:

.. code-block:: yaml

    apiVersion: v1
    kind: ConfigMap
    metadata:
      namespace: <namespace>
      name: <config-name>
    data:
      config.yaml: |-
        apiVersion: <object-version>
        kind: <kind>
        spec:
          <key>: <values>

**Use case 1:**

Configure and store the number of replicas for service specific Deployments
found in the `metalk8s-monitoring` namespace using the ConfigMap store format.

.. code-block:: yaml

    apiVersion: v1
    kind: ConfigMap
    metadata:
      namespace: metalk8s-monitoring
      name: metalk8s-grafana-userconfig
    data:
      config.yaml: |-
        apiVersion: metalk8s.scality.com/v1alpha1
        kind: GrafanaUserConfig
        spec:
          replicas: 2
        podSelector:

>           app: grafana

Reviewer: Probably missing an indent?

How it works
~~~~~~~~~~~~

During Bootstrap, pillar values for the various services will be pre-filled
with default values. Service pods and deployments will be configured to either

> consume configuration data directly from their ConfigMap store or fallback to

Reviewer: Using simply _ConfigMap_ i.s.o "ConfigMap store" or "ConfigMap KV store" should be sufficient for readers - could you change this here and in other relevant places?

the default pillar values if the service specific ConfigMap store was not found.

> **Using Saltstates**

Reviewer: Let's write this "Salt states" and use the same notation everywhere (no "Saltstate", "salt state", or "salt-state")

> Once a ConfigMap KV store is updated by the user(say a user changes the

Reviewer: ```suggestion
Once a ConfigMap KV store is updated by the user (say a user changes the
```

number of replicas for Prometheus deployments to a new value), then perform the

> following actions;

Reviewer: ```suggestion
following actions:
```

  - Apply a salt state that reads the ConfigMap object, validates the schema
    and checks the new values passed.

  - Finally, we make sure that the updated values are picked up by their

>     respective consumers(this might require Pod restarts for changes to take

Reviewer: ```suggestion
    respective consumers (this might require Pod restarts for changes to take
```

>     effect). The podSelector field has labels which can be used to detect which

Reviewer: There should be no need for us to "detect" what needs to restart: we own these objects, and the configuration options we expose, so it's safe to assume we'll always know what to do after a re-configuration state is applied.

    deployments to restart.

Note that, salt-states are used to sync data and update consumers
of new configurations changes mainly because of the minimum effort it takes to
setup this flow but the K8s Operator pattern could be use to replace
configuration synchronization between user defined configurations and
consumers.

The Operator approach is much more complex, requires much more effort
to realize and there is no real need for applying changes using this method
because configuration changes are not frequent(for a typical MetalK8s admin,
changing the number of replicas for a given deployment could happen once in 3
months) as such, having an operator watch for object changes is not significant

> and not very useful at this point in time.

Reviewer: Agreed with your points. Would just add that Operators are especially useful to provide self-healing functionalities on a reactive basis. When a user changes configuration, it is easy to apply (in the same flow) those configurations to the in-cluster objects. Once deployed, those configurations will be handled by the cluster controllers (Deployments, ConfigMap/Secret-based volume mounts, ...), and self-healing comes out of the box. No real need for more than what Salt provides.

> **Using Operator architecture(Custom Controllers)**

Reviewer: No need to provide a design if it's rejected. Simply explain why it's rejected, this will help readers quickly understand the overall architecture document.

When using an Operator(a Custom Controller that works with CRDs), we create a
Custom Resource Definition (CRD) that references a ConfigMap. Once a ConfigMap
is updated by the user, then the Operator is designed to perform the following
actions;

  - The Operator is connected to the API server to watch for changes in the
    ConfigMap.
  - If the Operator detects a modified ConfigMap, it then determines which line
    of action it should take which are;

      - Extract the ConfigMap name and object fields
      - Extract the pods associated to this ConfigMap based on it's labels
      - Read and validate the ConfigMap data and schema
      - If the ConfigMap is valid, update the respective pods such that they
        pick-up new configs.
      - If the ConfigMap is invalid, log the error and perform no further
        action.

Iteration 1
~~~~~~~~~~~

- Define and deploy new ConfigMap stores that will hold user configurations
  as listed in the requirements
- Template and render Deployment and Pod manifests that will make use of

>   this user customizable configurations using pillar values

Reviewer: Are we using Pillar values or directly call into `metalk8s_kubernetes` module in the end? cc. @TeddyAndrieux 

Maybe this could be a good fit for `__slots__` (and especially with [dict support in Salt 3000](https://docs.saltstack.com/en/latest/topics/slots/index.html))?

Reviewer 3: IMO we should use directly `metalk8s_kubernetes` module.

I think we don't really need `__slots__` here, basic jinja is enough

Reviewer 2: I'd try to steer away from `__slots__` as well. The values for these settings can be calculated/retrieved before templating starts, so let's do that.

- Document how to change user configurations using kubectl
- Create and deploy persistent storage volumes for Grafana dashboards and

>   datasources

Reviewer: We need to configure Grafana to use a PVC (not create the volumes ourselves). Just do the same as Prom.
Also, if it's out of the design (and I think we have an epic for Grafana: #2218), maybe just don't state it here.

- Document how to create these persistent volume for Grafana dashboards and
  datasources

> - Document how to add Prometheus alerting rules using a ConfigMap. This

Reviewer: Why not use the existing `PrometheusRule` CRD that comes with `prometheus-operator`? If it's a conscious decision, it would make sense to explain in the design choices.

Author: Will keep it simple and resolve to users editing the original CRD!

  documentation should explain to users that the Prometheus object has a
  ruleSelector field, which selects the ConfigMaps from which rules files are
  loaded from. This means that to add new Prometheus alert rules, Users need to
  create new ConfigMaps containing alert rules with labels matching what was

>   defined in the Prometheus ruleSelector field.

Reviewer: Ah! Actually, this is true if you replace `ConfigMap` with `PrometheusRule`, IIUC. Or maybe not in our version of `prometheus-operator`...? Didn't check

Author: Not tested, so let's leave it at editing the original CRD.

Iteration 2
~~~~~~~~~~~

- Provide a CLI tool for changing any of the user configurations:

>     - Count of replicas for chosen Deployments(Prometheus)

Reviewer: General comment: add a space before opening a parenthesis.

```suggestion
    - Count of replicas for chosen Deployments (Prometheus)
```

>     - Updating a Dex authentication connector(OpenLDAP, AD and

Reviewer: ```suggestion
    - Updating a Dex authentication connector (OpenLDAP, AD and
```

      staticUser store)
    - Updating the Alertmanager notification configuration

- Provide a UI interface that allows Update operations on all user customizable

>   settings based on the above requirements

Reviewer: From the requirements, CLI provides this "all settings possible" capability, while GUI provides more user-friendly / less detailed forms to configure certain services (as you described below).
I think (need to be confirmed) this point can be removed.

- Provide a UI interface for adding, updating and deleting service specific
  configurations for example Dex-LDAP  connector integration.
- Provide a UI interface for listing MetalK8s available/supported Dex
  authentication Connectors
- Provide a UI interface for enabling or disabling Dex authentication
  connectors(LDAP, Active Directory, StaticUser store)
- Provide a UI interface for changing the number of replicas for a chosen set
  of MetalK8s deployments(Prometheus, ....)
- Add a UI interface for listing Alertmanager notification systems MetalK8s
  will support(Slack, email, hipchat)
- Provide a UI interface for adding, modifying and deleting Alertmanager
  configurations from the listing above

Documentation
-------------

In the Operational Guide:

* Document how to customize or change any given service settings using the CLI
  tool
* Document how to customize or change any given service settings using the UI
  interface

> * Document how to add and edit Prometheus alerting rules using the CLI

Reviewer: Why this specific example and not all services?

Test Plan
---------

- Dex Static User authentication is currently covered in our test-suite and it
  will make sense to cover atleast one other authentication connector with the
  easiest being LDAP since we readily have access to OpenLDAP Docker images and

>   automating this process is possible on Scality Cloud

Reviewer: Out of scope - this relates to testing Dex, which is part of the Authentication document.

Configure and store the number of replicas for service specific Deployments
found in the `metalk8s-monitoring` namespace using the ConfigMap store format.

.. code-block:: yaml

    apiVersion: v1
    kind: ConfigMap
    metadata:
      namespace: metalk8s-monitoring
      name: metalk8s-grafana-userconfig
    data:
      config.yaml: |-
        apiVersion: metalk8s.scality.com/v1alpha1
        kind: GrafanaUserConfig
        spec:
          replicas: 2

How it works
~~~~~~~~~~~~

Service pods and deployments will be configured to consume configuration data
directly from their respective minion external pillars.

During Bootstrap, these external pillar values will be pre-filled with default
values and the service consumers will be configured to use these values.

**Using Saltstates**

Once a ConfigMap KV store is updated by the user(say a user changes the
number of replicas for Prometheus deployments to a new value), then perform the
following actions;

  - Apply a salt state that reads the ConfigMap object, validates the schema
    based on MetalK8s defined standards and checks the new values passed.
  - If the ConfigMap object is valid, the new values passed by the user are

>     then re-rendered to the pillars.

Reviewer 3: Not sure we really want this value to be in the pillar, IMO just reading the CM directly in the state is better so that user can NOT change, for example, the number of replicas just on CLI using `pillar="blabla"` but he always need to edit the CM

Author: Fixed

  - Finally, we make sure that the updated values are picked up by their
    respective consumers(this might require Pod restarts for changes to take
    effect).

Note that, salt-states are used to sync data and update consumers
of new configurations changes mainly because of the minimum effort it takes to
setup this flow(i.e. the act of configuration update by the user and it's
propagation to the consumers) but the K8s Operator pattern could be use to
replace configuration synchronization between user defined configurations and

> consumers.

Reviewer 3: Not sure we should talk about it here, for me we should take a desicion in this document and maybe do not implement the operator in the first iteration and only use salt but at least having the final view here is better IMO

The Operator approach is much more complex, requires much more effort
to realize and there is no real need for applying changes using this method
because configuration changes are not frequent(for a typical MetalK8s admin,
changing the number of replicas for a given deployment could happen once in 3
months) as such, having an operator watch for object changes is not significant
and not very useful at this point in time.

**Using Operator architecture(Custom Controllers)**

When using an Operator(a Custom Controller that works with CRDs), we create a
Custom Resource Definition (CRD) that references a ConfigMap. Once a ConfigMap
is updated by the user, then the Operator is designed to perform the following

> actions;

Reviewer 3: Operator and CRD are not related we could use CRD with salt and use Operator with ConfigMap.

Also if we use Operator and CRD then we no longer need ConfigMap IMO we directly use CRD as entry point

  - The Operator is connected to the API server to watch for changes in the
    ConfigMap.
  - If the Operator detects a modified ConfigMap, it then determines which line
    of action it should take which are;

      - Extract the ConfigMap name and object fields
      - Extract the pods associated to this ConfigMap based on it's labels
      - Read and validate the ConfigMap data and schema
      - If the ConfigMap is valid, update the pillars and restart the

>         respective pods such that they pick-up new configs from the pillars.

Reviewer 3: Not from the pillar in this approach

Author: As discussed in person... very true!

      - If the ConfigMap is invalid, log the error and perform no further
        action because a bad ConfigMap being applied could lead to cluster

>         outages

Reviewer 3: Hmm I not really like this point, we not directly apply the ConfigMap but we use it to modify another object and i don't see case where an invalid value can cause outages for me apiServer will just not set the wrong value.
For example if you use a string as replica then opeator will try to edit an object and apiServer will just reject the change, ... 
Maybe for other custom tuning that could lead to cluster outages but ... I don't see where

Author: ok... resolved!
